README.rst
django-turnapi
=======================
TURN API for django

*TURN Server*

https://code.google.com/p/rfc5766-turn-server/

*TURN Server API Proposal*

http://tools.ietf.org/html/draft-uberti-rtcweb-turn-rest-00

*Old TURN REST API on Google Docs*

https://docs.google.com/document/d/1mG7eXFQ5o-ypMWQ1IzdkBQL0UBkLN1xXUJhJcIF5ujQ

------------
Installation
------------
python setup.py install

----------
How to USE
----------

1. Configuration::

    TURN_REDIS_HOST = 'localhost'
    TURN_REDIS_PORT = 6379
    TURN_REDIS_DB   = 0 
    TURN_REDIS_PASSWORD = None
    TURN_REDIS_UNIX_DOMAIN_SOCKET_PATH = None
    TURN_REDIS_URL = None

    TURN_CREDENTIAS_TIMEOUT = 300                                                                                                                            
    TURN_API_URLS = []
    TURN_SHARED_SECRET = 'mySecrete'
    TURN_REALM = "mystite.com"
    TURN_SEPARATOR = ":"
    TURN_AUTH = True

2. Add turn api to your app settings::

    from turnapi import sites

    url(r'^turn$', include(sites.urls()) ),

-------
License
-------
MIT


CHANGELOG.md
0.0.2 (26-11-2013)
-----------------

  Add Shared Key Updater

0.0.1 (04-06-2013)
------------------

 _Very first, initial release_.


turnapi/api.py
"""api.py: Implemente rfc5766-turn-server HTTP server """

from django.http import HttpResponse
from django.utils import simplejson

import time
import hmac
from hashlib import sha1 as  hash_alg
import base64

from turnapi import settings

import redis

import logging

# REST API or dynamic Long term credentials
if settings.TURN_AUTH:
  from rfc5766_turn_server_auth import calc_key
else:
  from rfc5766_turn_server import calc_key

logger = logging.getLogger( 'turnapi.api' )

if settings.TURN_REDIS_URL is not None:
  redis_server = redis.StrictRedis.from_url( settings.TURN_REDIS_URL )
elif settings.TURN_REDIS_UNIX_DOMAIN_SOCKET_PATH is None:
  redis_con = redis.StrictRedis(
    host     = settings.TURN_REDIS_HOST,
    port     = settings.TURN_REDIS_PORT,
    db       = settings.TURN_REDIS_DB,
    password = settings.TURN_REDIS_PASSWORD
  )
else:
  redis_server = redis.StrictRedis(
    unix_socket_path = settings.TURN_REDIS_UNIX_DOMAIN_SOCKET_PATH,
    db               = settings.TURN_REDIS_DB,
    password         = settings.TURN_REDIS_PASSWORD,
  )

realm = settings.TURN_REALM

def turn( req ):
  query = req.GET

  service  = query.get( 'service',  None )
  username = query.get( 'username', None )
  ttl      = query.get( 'ttl'     , None )

  # Move this configurations to anywhere else!
  # SHared Secret
  shared_secret = settings.TURN_SHARED_SECRET

  # Get request timestamp (Seconds since 1970)
  timestamp = str( int( time.time() ) + (ttl if ttl else 86400) )

  # Username is the paramter plus timestamp with an separator
  # if username not defined just use timestamp as username
  username = str( timestamp + settings.TURN_SEPARATOR + username if username else timestamp )

  # Call key
  temp_pass = calc_key( username, realm, shared_secret )

  if not settings.TURN_AUTH:
    pKey = 'turn/user/%s/key' % ( username )
    kto  = settings.TURN_CREDENTIAS_TIMEOUT # seconds

    # Store credentials into Redis
    redis_con.setex( pKey, kto, temp_pass )

    # return plain text pass
    temp_pass = shared_secret

  items = {
    # "username" : "foo:" + timestamp,
    "username" : username,
    "password" : temp_pass,
    "ttl" : ttl if ttl else 86400,
    "uris" : settings.TURN_API_URLS
  }

  logger.debug( 'Response: ' + str(items) );

  # Send JSON response
  items = simplejson.dumps( items )
  response = HttpResponse( items, content_type = 'application/json; charset=utf8' )

  return response


turnapi/sites.py
# from functools import update_wrapper
# from django.http import Http404, HttpResponseRedirect
# from django.contrib.admin import ModelAdmin, actions
# from django.contrib.admin.forms import AdminAuthenticationForm
# from django.contrib.auth import REDIRECT_FIELD_NAME
# from django.contrib.contenttypes import views as contenttype_views
# from django.views.decorators.csrf import csrf_protect
# from django.db.models.base import ModelBase
# from django.core.exceptions import ImproperlyConfigured
# from django.core.urlresolvers import reverse, NoReverseMatch
# from django.template.response import TemplateResponse
# from django.utils import six
# from django.utils.text import capfirst
# from django.utils.translation import ugettext as _
# from django.views.decorators.cache import never_cache
from django.conf import settings

from django.conf.urls import patterns, url, include

# LOGIN_FORM_KEY = 'this_is_the_login_form'

# class AlreadyRegistered(Exception):
#     pass
# 
# 
# class NotRegistered(Exception):
#     pass


def get_urls():
  # if settings.DEBUG:
  #  self.check_dependencies()

  # Admin-site-wide views.
  urlpatterns = patterns('',
    url(r'^$',
      'turnapi.api.turn', # wrap(self.index),
      name='turnapi')
  )

  # # Add in each model's views.
  # for model, model_admin in six.iteritems(self._registry):
  #   urlpatterns += patterns('',
  #     url(r'^%s/%s/' % (model._meta.app_label, model._meta.module_name),
  #       include(model_admin.urls))
  #   )
  return urlpatterns

def urls():
  return get_urls(), 'self.app_name', 'self.name'


turnapi/shared_key_updater.py
""" Implemente Shared Key Updater """

from turnapi import settings

import redis

from threading import Timer
import time

import logging

logger = logging.getLogger( 'turnapi.api' )

to = settings.TURN_KEY_TTL

from datetime import datetime 

keyUpdater = None

# Keys Updater
def genKey( data, first = False ):
  logger.debug( "Gen new key %s " % ( datetime.now() ) )

  shared_secret = data['shared secret']
  redis_con = data['redis connection']

  logger.debug( "Shared Secret (Old): %s" % ( shared_secret ) )

  # TODO: Gen random secret
  # Mod 10 to keep number lengt equal to 1 xD
  index = ( int(shared_secret[-1:] ) + 1 ) % 10
  data['shared secret'] = shared_secret[:-1] + str( index )

  # the keys are always synchronized on timestamp
  ts = int( time.time() )
  ts = int(ts / to) * to

  # if the first gen a key behind otherwise
  # generates the next one
  if not first:
    ts += to

  logger.debug( "Set key %d => %s " % ( ts, datetime.fromtimestamp( ts ) ) )
  key = "turn/secret/%d" % ( ts )
  ret = redis_con.setnx( key, shared_secret )
  if ret == 1:
    logger.debug( "Set Key TTL %d, expire at %s" % (to, datetime.fromtimestamp( ts + to ) ) )
    redis_con.expireat( key, ts + to )
  else:
    data['shared secret'] = redis_con.get( key )

  logger.debug( "Shared Secret (New): %s" % ( shared_secret ) )

  if first:
    return

  # Schedule next update
  nxt = datetime.fromtimestamp( ts + to * 2 / 3 )
  now = datetime.now()
  waitTime = (nxt - now).total_seconds()
  if waitTime <= 0:
    waitTime = 1

  logger.debug( "Next Key at %s (%d)" % ( datetime.fromtimestamp( time.time() + waitTime), waitTime ) )

  keyUpdater = Timer( waitTime, genKey, (data, False) )
  keyUpdater.start()
  return
# end genKey

def Start( data ):
  genKey( data )
  # generates the previous key if not define
  genKey( data, True )
# end Start

def Stop():
  logger.warning( "Not Imlemented" )
# end Stop


turnapi/settings.py
from django.conf import settings

TURN_REDIS_HOST = getattr(settings, 'TURN_REDIS_HOST', 'localhost')
TURN_REDIS_PORT = getattr(settings, 'TURN_REDIS_PORT', 6379)
TURN_REDIS_DB = getattr(settings, 'TURN_REDIS_DB', 0)
TURN_REDIS_PASSWORD = getattr(settings, 'TURN_REDIS_PASSWORD', None)
TURN_REDIS_UNIX_DOMAIN_SOCKET_PATH = getattr(
    settings, 'TURN_REDIS_UNIX_DOMAIN_SOCKET_PATH', None
)
TURN_REDIS_URL = getattr(settings, 'TURN_REDIS_URL', None)

TURN_CREDENTIAS_TIMEOUT = getattr(settings, 'TURN_CREDENTIAS_TIMEOUT', 300)
TURN_API_URLS = getattr(settings, 'TURN_API_URLS', [])
TURN_SHARED_SECRET = getattr(settings, 'TURN_SHARED_SECRET', 'mySecrete')
TURN_REALM = getattr(settings, 'TURN_REALM', 'mysite.com')
TURN_SEPARATOR = getattr(settings, 'TURN_SEPARATOR', ':')
TURN_AUTH = getattr(settings, 'TURN_AUTH', True)

TURN_KEY_TTL = getattr(settings, 'TURN_KEY_TTL', 86400)
